### PR TITLE
feat: update type signature during add argument action

### PR DESCRIPTION
**Pull request by santiweight**
@michaelpj @pepeiborra This one resolves a lot of the issues you had with the previous MR

It's also hopefully much clearer now, since I spent some time cleaning up also...

Some notes:
 - A future MR will split the logic into a separate file (I am delaying due to merge conflicts that I want to avoid)
 - There is still one uncompleted feature that is noted in documentation (we do not use GHC's suggested type when updating the type signature)

<a href="https://gitpod.io/#https://github.com/haskell/haskell-language-server/pull/3321"></a>

### CODE REVIEW

**Comment by santiweight** [comment]
One point of annoyance I had during implementation: you can't run golden tests without setting `-j1`. Running golden tests multi-threaded would have bizarre behavior around the current working directory and creating golden files in odd places.

It took me a while to realize... is there a way to fix `runSessionWithServer'` so that future contributors are not burned by this?

File: plugins/hls-refactor-plugin/src/Development/IDE/Plugin/CodeAction.hs [hunk 1]
```diff
@@ -972,31 +982,97 @@ suggestAddArgument parsedModule Diagnostic {_message, _range}
...
+
+-- Attempt to insert a binding pattern into each match for the given LHsDecl. Also return the declaration's name, and
+-- the number of bound patterns in the declaration's matches, if the HsDecl is a fun bind
+--
+--    insertArg "new_pat" `foo bar baz = 1`
+-- => (`foo bar baz new_pat = 1`, Just ("foo", 2))
+appendFinalPatToMatches :: T.Text -> LHsDecl GhcPs -> TransformT (Either ResponseError) (LHsDecl GhcPs, Maybe (GenLocated SrcSpanAnnN RdrName, Int))
+appendFinalPatToMatches name = \case
+  (L locDecl (ValD xVal (FunBind xFunBind idFunBind mg coreFunBind))) -> do
+    (mg', numPatsMay) <- modifyMgMatchesT' mg (pure . second Just . addArgToMatch name) Nothing combineMatchNumPats
```
Comment (michaelpj) [inline plugins/hls-refactor-plugin/src/Development/IDE/Plugin/CodeAction.hs:1003]: This all feels a bit roundabout. You've taught two functions to pass up the number of matches, and then changed `modifyMgMatches` to pass a result thorough... but it seems like you could just _look_ for the number of matches independently of the modification, either before or afterwards. Sure, you have to write another function, but I imagine `funBindArgCount` is quite straightforward code and simplifies all the rest.

File: plugins/hls-refactor-plugin/test/data/golden/add-arg/AddArgWithTypeSynSig.hs [hunk 2]
```diff
@@ -0,0 +1,5 @@
+type FunctionTySyn = Bool -> Int
```
Comment (michaelpj) [inline plugins/hls-refactor-plugin/test/data/golden/add-arg/AddArgWithTypeSynSig.hs:1]: great edge case

**Review by michaelpj — APPROVED** [review]
Somewhat superficial review, but looks generally good! Not sure about the tests.

File: plugins/hls-refactor-plugin/src/Development/IDE/GHC/ExactPrint.hs [hunk 3]
```diff
@@ -450,31 +460,125 @@ graftDecls dst decs0 = Graft $ \dflags a -> do
...
 modifySmallestDeclWithM ::
-  forall a m.
+  forall a m r.
   (HasDecls a, Monad m) =>
   (SrcSpan -> m Bool) ->
-  (LHsDecl GhcPs -> TransformT m [LHsDecl GhcPs]) ->
+  (LHsDecl GhcPs -> TransformT m ([LHsDecl GhcPs], r)) ->
   a ->
-  TransformT m a
+  TransformT m (a, Maybe r)
```
[the remark this answers is not in the document]
Comment (santiweight) [reply plugins/hls-refactor-plugin/src/Development/IDE/GHC/ExactPrint.hs:468]: Revised

File: plugins/hls-refactor-plugin/src/Development/IDE/GHC/ExactPrint.hs [hunk 4]
```diff
@@ -450,31 +460,125 @@ graftDecls dst decs0 = Graft $ \dflags a -> do
...
+setAnchor _ spanAnnN = spanAnnN
+
+removeTrailingAnns :: SrcSpanAnnN -> SrcSpanAnnN
+removeTrailingAnns (SrcSpanAnn (EpAnn anc nameAnn comments) span) =
+  let nameAnnSansTrailings = nameAnn {nann_trailing = []}
+  in SrcSpanAnn (EpAnn anc nameAnnSansTrailings comments) span
+removeTrailingAnns spanAnnN = spanAnnN
+
+-- | Modify the type signature for the given IdP. This function handles splitting a multi-sig
+-- SigD into multiple SigD if the type signature is changed.
```
[the remark this answers is not in the document]
Comment (santiweight) [reply plugins/hls-refactor-plugin/src/Development/IDE/GHC/ExactPrint.hs:494]: Oddly enough this took less than most of the simple stuff 🤷‍♂️

File: plugins/hls-refactor-plugin/src/Development/IDE/GHC/ExactPrint.hs [hunk 5]
```diff
@@ -450,31 +460,125 @@ graftDecls dst decs0 = Graft $ \dflags a -> do
...
   Monad m =>
   MatchGroup GhcPs (LHsExpr GhcPs) ->
   (LMatch GhcPs (LHsExpr GhcPs) -> TransformT m (LMatch GhcPs (LHsExpr GhcPs))) ->
   TransformT m (MatchGroup GhcPs (LHsExpr GhcPs))
-modifyMgMatchesT (MG xMg (L locMatches matches) originMg) f = do
-  matches' <- mapM f matches
-  pure $ MG xMg (L locMatches matches') originMg
+modifyMgMatchesT mg f = fst <$> modifyMgMatchesT' mg (fmap (, ()) . f) () ((.) pure . const)
+
+-- | Modify the each LMatch in a MatchGroup
```
[the remark this answers is not in the document]
Comment (santiweight) [reply plugins/hls-refactor-plugin/src/Development/IDE/GHC/ExactPrint.hs:570]: Revised

File: plugins/hls-refactor-plugin/src/Development/IDE/Plugin/CodeAction.hs [hunk 6]
```diff
@@ -972,31 +982,97 @@ suggestAddArgument parsedModule Diagnostic {_message, _range}
   where
     message = unifySpaces _message
 
--- TODO use typ to modify type signature
+-- Given a name for the new binding, add a new pattern to the match in the last position,
+-- returning how many patterns are in this match:
```
[the remark this answers is not in the document]
Comment (santiweight) [reply plugins/hls-refactor-plugin/src/Development/IDE/Plugin/CodeAction.hs:986]: Before. Revised

File: plugins/hls-refactor-plugin/src/Development/IDE/Plugin/CodeAction.hs [hunk 7]
```diff
@@ -972,31 +982,97 @@ suggestAddArgument parsedModule Diagnostic {_message, _range}
...
+--      addArgToMatch "foo" `bar arg1 arg2 = ...`
+--   => (`bar arg1 arg2 foo = ...`, 2)
+addArgToMatch :: T.Text -> GenLocated l (Match GhcPs body) -> (GenLocated l (Match GhcPs body), Int)
+addArgToMatch name (L locMatch (Match xMatch ctxMatch pats rhs)) =
+  let unqualName = mkRdrUnqual $ mkVarOcc $ T.unpack name
+      newPat = L (noAnnSrcSpanDP1 generatedSrcSpan) $ VarPat NoExtField (noLocA unqualName)
+  in (L locMatch (Match xMatch ctxMatch (pats <> [newPat]) rhs), length pats)
+
+-- Attempt to insert a binding pattern into each match for the given LHsDecl. Also return the declaration's name, and
+-- the number of bound patterns in the declaration's matches, if the HsDecl is a fun bind
```
[the remark this answers is not in the document]
Comment (santiweight) [reply plugins/hls-refactor-plugin/src/Development/IDE/Plugin/CodeAction.hs:996]: Revised

File: plugins/hls-refactor-plugin/src/Development/IDE/Plugin/CodeAction.hs (hunk 1 above)
Comment (santiweight) [reply plugins/hls-refactor-plugin/src/Development/IDE/Plugin/CodeAction.hs:1003]: The unfortunate thing is that we don't know which declaration is going to be altered.

Note that we use `modifySmallestDeclWithM` with this function in `addArgumentAction`.

An alternative would be to do two calls to `modifySmallestDeclWithM` (perhaps a variant called `viewSmallestDeclWithM` that just returns some information on the matching decl). The first would do the modification, the second would extract the information (and then the two maybes would be `<|>` together).

I don't personally find that clearer... While I do agree that this code is a little odd, I think it's part and parcel of these sorts of transformations :/

I'm happy to switch it up if you still think it'd be clearer.

Comment (michaelpj) [reply plugins/hls-refactor-plugin/src/Development/IDE/Plugin/CodeAction.hs:1006]: Okay, so you do quite a bit of finding this decl. You have to go in and find it here to modify it, then you have to go in again to find the matching signature?

Continuing possibly dumb questions: could we not do something like:
```
do
   decl <- findSmallestDecl
   sig <- findSigFor decl
   -- return the modified version in case we want to look at them
   decl' <- modifyDecl decl
   sig' <- modifySig sig
```
Anyway, I won't nitpick more: I'm happy to go with what you think makes sense as a structure here. It just feels like we've got some pieces of work that are currently entangled and I wonder if they could be disentangled.

**Comment by michaelpj** [comment]
What happens if you run this on a local def in a let or where? does it work? Can we have some tests to show whatever happens? And what happens if there's an intervening local def like the lambda case?

File: plugins/hls-refactor-plugin/src/Development/IDE/Plugin/CodeAction.hs (hunk 1 above)
Comment (santiweight) [reply plugins/hls-refactor-plugin/src/Development/IDE/Plugin/CodeAction.hs:1007]: Ah I see your thought. We certainly could do something like this.

However, this not generally how I've seen other AST refactors done with ghc-exactprint. Normally I see modifications done in a lens style `Lens ast Decl`, and I think it makes some sense to continue to do so.

The API you suggest is actually something I've been considering myself for a while, but the exactprint API is far from stable enough to have such niceties (there would probably be a fair amount of CPP involved in the near future...).

**Comment by santiweight** [comment]
@michaelpj 

There were already tests for where bindings in the tests that existed before support for updating type signatures. However, I've added golden tests because the old tests were crap (I will be refactoring them in a follow up).

This plugin currently only looks for top-level bindings.

```Haskell
foo = bar
  where bar = new_def
-- Hover over new_def...

-- current behavior (only for top levels): 
+ add as argument

-- future behavior:
+ add as argument to foo
+ add as argument to bar
```
